glean-core/csharp/GleanTests/Metrics/PingTypeTest.cs

```diff
@@ -0,0 +1,91 @@
+﻿using System;
+using System.IO;
+using Xunit;
+using static Mozilla.Glean.Glean;
+
+namespace Mozilla.Glean.Tests.Metrics
+{
+    public class PingTypeTest
+    {
+        public PingTypeTest()
+        {
+            // Get a random test directory just for this single test.
+            string tempDataDir = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
+
+            // In xUnit, the constructor will be called before each test. This
+            // feels like a natural place to initialize / reset Glean.
+            GleanInstance.Reset(
+                applicationId: "org.mozilla.csharp.tests",
+                applicationVersion: "1.0-test",
+                uploadEnabled: true,
+                configuration: new Configuration(),
+                dataDir: tempDataDir
+                );
+        }
+
+        [Fact]
+        public void SendCustomPings()
+        {
+            Private.PingType<Private.NoReasonCodes> customPing =
+                new Private.PingType<Private.NoReasonCodes>(
+                    name: "custom",
+                    includeClientId: true,
+                    sendIfEmpty: false,
+                    reasonCodes: null
+                );
+
+            Assert.True(GleanInstance.TestHasPingType("custom"));
+
+            // TODO: Using CounterMetricType to test custom pings
+            // should make more sense.
+            //Private.BooleanMetricType booleanMetric = new Private.BooleanMetricType(
+            //    category: "telemetry",
+            //    disabled: false,
+            //    lifetime: Private.Lifetime.Application,
+            //    name: "boolean_metric",
+            //    sendInPings: new string[] { "custom" }
+            //);
+
+            //// Record two strings of the same type, with a little delay.
+            //booleanMetric.Set(true);
+
+            //// Check that data was properly recorded.
+            //Assert.True(booleanMetric.TestHasValue());
+            customPing.Submit();
+        }
+
+        [Fact]
+        public void SendCustomPingsWithSnakeCase()
+        {
+            // TODO:
+        }
+
+        [Fact]
+        public void SendCustomPingsWithKebabCase()
+        {
+            // TODO:
+        }
+
+        [Fact]
+        public void SendCustomPingsWithoutClientId()
+        {
+            // TODO:
+        }
+
+        [Fact]
+        public void SendCustomPingsWithAnUnknownNameNoOp()
+        {
+            // TODO:
+        }
+
+        [Fact]
+        public void RegistryShouldContainBuildInPings()
+        {
+            // Blow lines return false always, how to I register
+            // build-in pings?
+            //Assert.True(GleanInstance.TestHasPingType("metrics"));
```

Comment:
If I register pings and their names are `metrics`, `baseline`, `event` before calling below lines. These tests will return true. However, I feel these three types pings are build-in pings. We should register them internally, but I don't know where. @Dexterp37

Reply:
Good point. The pings code is usually generated automatically by the build system, using the `glean_parser`, when building the Glean SDK. Unfortunately, we don't have that part yet, so we need to generate a file manually, as we do for our [core metrics](https://searchfox.org/glean/source/glean-core/csharp/Glean/ManualMetrics.cs). You need to create a `ManualPings.cs` that defines the pings.

For example, [here's the content](https://dpaste.org/RP2o) of the `Pings.kt` file generated by the Glean build system when parsing the SDK.

Ping registration can then happen [here](https://searchfox.org/glean/rev/bf1825a7523576911c46eb770912630693228a0e/glean-core/csharp/Glean/Glean.cs#92), by passing in the manual pings class.

Reply:
I have made a `ManualPings.cs` and was trying to lazy create these `PingTypes` as what you did in `ManualMetrics.cs`. However, when executing running all tests, we will call `Glean.TestDestroyGleanHandle()` to destroy Glean between tests. That makes me have to re-create these build-in ping types after every time we initialize Glean. For the case of running a single test, we can just lazy create them.